### PR TITLE
ENH: Refactor DataDifferencer to reduce duplicate code.

### PR DESCRIPTION
**Pull request by GuySartorelli**
Also remove cloning where it isn't necessary.

@michalkleiner per our discussion starting in [this comment](https://github.com/silverstripe/silverstripe-versioned/pull/354#issuecomment-1068788809) on #354.

I didn't make a separate method in the end because the list of diffed fields is exactly what is returned by `changedFieldNames`.

### CODE REVIEW

File: src/DataDifferencer.php [hunk 1]
```diff
@@ -208,45 +208,32 @@ protected function getObjectDisplay($object = null)
      */
     public function ChangedFields()
     {
+        $base = $this->fromRecord ?: $this->toRecord;
         $changedFields = new ArrayList();
```
Comment (michalkleiner) [inline src/DataDifferencer.php:212]:
```suggestion
        $changedFields = ArrayList::create();
```

File: src/DataDifferencer.php [hunk 2]
```diff
@@ -208,45 +208,32 @@ protected function getObjectDisplay($object = null)
...
                 }
-                $changedFields->push(new ArrayData([
-                    'Name' => $field,
-                    'Title' => $base->fieldLabel($field),
-                    'Diff' => $fieldDiff,
-                    'From' => $this->fromRecord ? $this->fromRecord->$field : null,
-                    'To' => $this->toRecord ? $this->toRecord->$field : null,
-                ]));
             }
+            $changedFields->push(new ArrayData([
```
Comment (michalkleiner) [inline src/DataDifferencer.php:230]:
```suggestion
            $changedFields->push(ArrayData::create([
```

**Review by michalkleiner — COMMENTED** [review]
Nice, that looks tidier.
Could you also please update to use the `::create` syntax since we're touching this.

**Comment by GuySartorelli** [comment]
Yeah can do. I noticed that and only left it alone because one of the things I aim on doing is a global overhaul of replacing `new` with `create` in all the repositories. But I have no problem doing this one as part of this PR :)

**Comment by GuySartorelli** [comment]
I've replaced instances of the `new` keyword in this class with dependency injection for `Injectable` classes.

**Review by michalkleiner — APPROVED** [review]

**Pull request merged**